src/herder/test/UpgradesTests.cpp

```diff
@@ -1831,3 +1831,58 @@ TEST_CASE("upgrade invalid during ledger close", "[upgrades]")
     REQUIRE_THROWS(executeUpgrade(*app, makeTxCountUpgrade(0)));
     REQUIRE_THROWS(executeUpgrade(*app, makeBaseReserveUpgrade(0)));
 }
+
+void
+testUpgradeExpiration(std::chrono::seconds durationAfterUpgradeTime,
+                      bool shouldExpire)
+{
+    auto cfg = getTestConfig();
+    cfg.LEDGER_PROTOCOL_VERSION = 10;
+    cfg.TESTING_UPGRADE_DESIRED_FEE = 100;
+    cfg.TESTING_UPGRADE_MAX_TX_SET_SIZE = 50;
+    cfg.TESTING_UPGRADE_RESERVE = 100000000;
+    cfg.TESTING_UPGRADE_DATETIME = genesis(0, 0);
+
+    auto header = LedgerHeader{};
+
+    // make sure the network info is different than what's armed
+    header.ledgerVersion = cfg.LEDGER_PROTOCOL_VERSION - 1;
+    header.baseFee = cfg.TESTING_UPGRADE_DESIRED_FEE - 1;
+    header.baseReserve = cfg.TESTING_UPGRADE_RESERVE - 1;
+    header.maxTxSetSize = cfg.TESTING_UPGRADE_MAX_TX_SET_SIZE - 1;
+    header.scpValue.closeTime = VirtualClock::to_time_t(
+        cfg.TESTING_UPGRADE_DATETIME + durationAfterUpgradeTime);
+
+    bool updated = false;
+    auto upgrades = Upgrades{cfg}.removeUpgrades(
+        header.scpValue.upgrades.begin(), header.scpValue.upgrades.end(),
+        header.scpValue.closeTime, updated);
+
+    if (shouldExpire)
+    {
+        REQUIRE(updated);
+        REQUIRE(!upgrades.mProtocolVersion);
+        REQUIRE(!upgrades.mBaseFee);
+        REQUIRE(!upgrades.mMaxTxSize);
+        REQUIRE(!upgrades.mBaseReserve);
+    }
+    else
+    {
+        REQUIRE(!updated);
+        REQUIRE(upgrades.mProtocolVersion);
+        REQUIRE(upgrades.mBaseFee);
+        REQUIRE(upgrades.mMaxTxSize);
+        REQUIRE(upgrades.mBaseReserve);
+    }
+}
+
+TEST_CASE("remove expired upgrades", "[upgrades]")
```

Comment:
We prefer to use sections in this case: not because the setup is expensive but to avoid creating too many ultra small tests (when we run parallel tests, we split using test cases).
Note that using a section also allows (if you want) to move the helper function inside the test case (as a lambda), as to scope it better.